File: calGen.pl
#!/usr/bin/perl
use strict;
use warnings;

my $file_name = "ctrla";
my $input_path = shift;
my $output_path = "output.ics";

open my $input_file, '<', $input_path
    or die "Error opening $input_path - $!\n";

open my $output_file, '>', $output_path
    or die "Error creating $output_path - $!\n";


# ICS file header
print $output_file "BEGIN:VCALENDAR\n";
print $output_file "VERSION:2.0\n";
print $output_file "PRODID:-//Kevin Thai//reQuestCal 1.0//EN\n";

my $numevents = 0;

# Iterate through Quest text
while (<$input_file>) {
  # Start of course info for one specific course
  if (/(\w+ \d+) - ([\w\- ]+)/) {
    # Start of course components info block
    my %row = (
      course_code => $1,
      course_name => $2,
      class_number => "",
      section => "",
      component => "",
      days_times => "",
      room => "",
      instructor => "",
      start_end => ""
    );

    # Skip 12 lines
    <$input_file> foreach (1..12);

    # Entries in each row
    $_ = <$input_file>;
    until (/\t/) { # Catch all rows
      $numevents += 1;
      read_row($input_file, \%row, $_);
      create_event($output_file, \%row);

      # Get next line
      $_ = <$input_file>;
    }
  }
}

# End ICS file
print $output_file "END:VCALENDAR\n";

print "The total number of events created was $numevents.\n";




# SUBROUTINES

# Parse the equivalent of a row of course component info and update data struct
sub read_row {
  my $input_file = shift;
  my $row_ref = shift;
  my $next_line = shift;

  unless ($next_line =~ /\ /) {
    # Regular row
    $$row_ref{class_number} = $next_line;
    <$input_file>;
    $$row_ref{section} = <$input_file>;
    <$input_file>;
    $$row_ref{component} = <$input_file>;
  } else {
    # Add'l line of same component
    <$input_file>;
    <$input_file>;
  }
  <$input_file>;
  $$row_ref{days_times} = <$input_file>;
  <$input_file>;
  $$row_ref{room} = <$input_file>;
  <$input_file>;
  $$row_ref{instructor} = <$input_file>;
  <$input_file>;
  $$row_ref{start_end} = <$input_file>;

  # Remove newlines
  chomp $$row_ref{$_} foreach (keys %$row_ref);
}

# Process the info and create event
sub create_event {
  my $output_file = shift;
  my $row_ref = shift;

  # Summary
  my $summary;
  if ($$row_ref{component} eq "LEC") {
    $summary = "$$row_ref{course_code} - $$row_ref{course_name}";
  } elsif ($$row_ref{component} eq "TST") {
    $summary = "$$row_ref{course_code} Midterm";
  } else {
    $summary = "$$row_ref{course_code} $$row_ref{component}";
  }

  # Description
  my $desc = "Class Nbr: $$row_ref{class_number}";
  $desc .= "\\nSection: $$row_ref{section}";
  $desc .= "\\nInstructor: $$row_ref{instructor}";

  # Location
  my $location = join(' ', split(/\s+/, $$row_ref{room}) );

  # DTSTAMP
  # format is YYYMMDDThhmmssZ
  my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = gmtime();
  my $dtstamp = sprintf("%04d%02d%02dT%02d%02d%02dZ",
    $year+1900, $mon+1, $mday, $hour, $min, $sec);

  # Parse out Start and End dates
  my @out = $$row_ref{start_end} =~ /(\d+)\/(\d+)\/(\d+)/g;
  my $start_date = join('', ($out[2], $out[0], $out[1]));
  my $end_date   = join('', ($out[5], $out[3], $out[4]));

  # DTSTART and DTEND and RRULE
  my $recurring = ($start_date ne $end_date);
  my ($dtstart, $dtend, $rrule);
  if ($recurring) {
    # Regular recurring component
    ($dtstart, $dtend) = create_dtbounds($$row_ref{days_times}, $start_date, $end_date);
    $rrule             = create_rrule(   $$row_ref{days_times}, $start_date, $end_date);
  } else {
    # Midterm or Seminar or Lab
    ($dtstart, $dtend) = create_dtbounds($$row_ref{days_times}, $start_date, $end_date);
  }

  # UID
  my $uid = sprintf("%04d%02d%02d%02d%02d%02d", $year+1900, $mon+1, $mday, $hour, $min, $sec);;
  $uid .= sprintf("%d", int(rand(999999)));
  $uid .= "\@aruu.github.com";

  # Sample VEVENT format

  # BEGIN:VEVENT
  # DTSTART;TZID=America/Toronto:20150415T143000
  # DTEND;TZID=America/Toronto:20150415T170000
  # RRULE:FREQ=WEEKLY;UNTIL=20150520T183000Z;BYDAY=WE
  # DTSTAMP:20150415T063712Z
  # UID:a6elgdb5paod2eajn8uehscbe4@google.com
  # CREATED:20150415T063702Z
  # DESCRIPTION:asdf\nasdf\nasdf\nasf
  # LAST-MODIFIED:20150415T063702Z
  # LOCATION:asd  as
  # SEQUENCE:0
  # STATUS:CONFIRMED
  # SUMMARY:asdfasdf
  # TRANSP:OPAQUE
  # END:VEVENT

  # BEGIN:VEVENT
  # DTSTART:20150414T153000Z
  # DTEND:20150414T180000Z
  # DTSTAMP:20150414T023441Z
  # UID:i5hiorj4c5k5jieieo19p8fet0@google.com
  # CREATED:20150414T023411Z
  # DESCRIPTION:desc\nasdf\nasdf
  # LAST-MODIFIED:20150414T023411Z
  # LOCATION:place
  # SEQUENCE:0
  # STATUS:CONFIRMED
  # SUMMARY:title
  # TRANSP:OPAQUE
  # END:VEVENT

  # Print to .ICS file
  print $output_file "BEGIN:VEVENT\n";
  print $output_file "DTSTART;TZID=America/Toronto:$dtstart\n";
  print $output_file "DTEND;TZID=America/Toronto:$dtend\n";
  print $output_file "RRULE:$rrule\n" if $recurring;
  print $output_file "DTSTAMP:$dtstamp\n";
  print $output_file "UID:$uid\n";
  # print $output_file "CREATED:\n";
  print $output_file "DESCRIPTION:$desc\n";
  # print $output_file "LAST-MODIFIED:\n";
  print $output_file "LOCATION:$location\n";
  # print $output_file "SEQUENCE:0\n";
  # print $output_file "STATUS:CONFIRMED\n";
  print $output_file "SUMMARY:$summary\n";
  # print $output_file "TRANSP:OPAQUE\n";
  print $output_file "END:VEVENT\n";
}

sub create_rrule {
  my $days_times = shift;
  my $start_date = shift;
  my $end_date = shift;

  my $rrule = sprintf("FREQ=WEEKLY;UNTIL=%sT200000Z;BYDAY=", $end_date);

  $days_times =~ /(\w+)/;
  my @days_chars = split(//, $1);
  my $toccur = 0;
  foreach my $ch (@days_chars) {
  if ($ch eq 'M') {
    $rrule .= "MO,";
  } elsif ($ch eq 'T') {
    $rrule .= "TU," if $toccur;
    $toccur = 1;
  } elsif ($ch eq 'W') {
    $rrule .= "TU," if $toccur;
    $toccur = 0;
    $rrule .= "WE,";
  } elsif ($ch eq 'h') {
    $toccur = 0;
    $rrule .= "TH,";
  } elsif ($ch eq 'F') {
    $rrule .= "TU," if $toccur;
    $toccur = 0;
    $rrule .= "FR,";
  }
  }
  $rrule .= "TU," if $toccur;
  chop $rrule;

  return $rrule;
}

sub create_dtbounds {
  my $days_times = shift;
  my $start_date = shift;
  my $end_date = shift;

  if ($start_date ne $end_date) {
    # Determine date of first occurrence,
    #   start date given is the first day of classes, Monday
    $days_times =~ /(\w+)/;
    my @days_chars = split(//, $1);
    my $daystr = '';
    my $toccur = 0;
    foreach my $ch (@days_chars) {
    if ($ch eq 'M') {
      $daystr .= "0";
    } elsif ($ch eq 'T') {
      $daystr .= "1" if $toccur;
      $toccur = 1;
    } elsif ($ch eq 'W') {
      $daystr .= "1" if $toccur;
      $toccur = 0;
      $daystr .= "2";
    } elsif ($ch eq 'h') {
      $toccur = 0;
      $daystr .= "3";
    } elsif ($ch eq 'F') {
      $daystr .= "1" if $toccur;
      $toccur = 0;
      $daystr .= "4";
    }
    }
    $daystr .= "1" if $toccur;
    $start_date += substr($daystr, 0, 1);
  }

  my @t = $days_times =~ /(\d+):(\d+)(\w+)/g;
  my ($a, $b);
  $t[0] += 12 if ($t[2] eq "PM" and $t[0] != 12);
  $a = sprintf("%02d%02d00", $t[0], $t[1]);
  $t[3] += 12 if ($t[5] eq "PM" and $t[3] != 12);
  # Have events ending at HH:30 and HH:00
  $t[4] += 10;
  if ($t[4] == 60) {
    $t[3] += 1;
    $t[4] = 0;
  }
  $b = sprintf("%02d%02d00", $t[3], $t[4]);

  my $dtstart = sprintf("%sT%06d", $start_date, $a);
  my $dtend   = sprintf("%sT%06d", $start_date, $b);

  return ($dtstart, $dtend);
}

File: calGen.js
var uidCounter = 0;

function main() {
  var input = document.getElementById("quest_content").value;

  // Parse text
  var rows = parseText(input);

  // now we have the table information as they exist, with inferences
  // console.log(rows);
  // could merge the biweekly lab events into a recurring event here

  // Populate table
  var formatTable = document.getElementById("format_table");
  populateFormatTable(rows, formatTable);

  // ICS file header
  var output = "";
  output += "BEGIN:VCALENDAR\n";
  output += "VERSION:2.0\n";
  output += "PRODID:-//Kevin Thai//reQuestCal 1.5//EN\n";

  for (var i=0; i<rows.length; i++) {
    // console.log(rows[i]);
    if (rows[i]["Days & Times"] !== undefined && rows[i]["Days & Times"] !== "TBA" && rows[i]["Days & Times"] !== " ") {
      output += createEvent(rows[i]);
    }
  }

  // End ICS file
  output += "END:VCALENDAR\n";


  // Output for FullCalendar
  // dude this problem is annoying because ICS support is lacking ><
  // maybe use ical.js from mozilla later on
  document.getElementById("ics_content").value = output;
  var events = [];
  var icsEvents = output.match(/BEGIN:VEVENT[\s\S]*?END:VEVENT/g);
  for (var i=0; i<icsEvents.length; i++) {
    var title = icsEvents[i].match(/SUMMARY:(.*)/)[1];
    var start = icsEvents[i].match(/DTSTART;TZID=America\/Toronto:(.*)/)[1];
    var end = icsEvents[i].match(/DTEND;TZID=America\/Toronto:(.*)/)[1];
    var rrule = icsEvents[i].match(/RRULE:(.*)/);
    var temp = {
      title: title,
      start: start,
      end: end
    }
    if (rrule === null) {
      events.push(temp);
    } else {
      var utcStart = moment.tz(start, "America/Toronto").toISOString().replace(/[-:]/g, "").replace(/\.\d{3}/, "");
      var rStarts = "RRULE:".concat(rrule[1]).concat(";DTSTART=").concat(utcStart);
      rStarts = rrulestr(rStarts).all();
      var utcEnd = moment.tz(end, "America/Toronto").toISOString().replace(/[-:]/g, "").replace(/\.\d{3}/, "");
      var rEnds = "RRULE:".concat(rrule[1]).concat(";DTSTART=").concat(utcEnd);
      rEnds = rrulestr(rEnds).all();
      for (var j=0; j<rStarts.length; j++) {
        var temp2 = {
          title: temp.title,
          start: rStarts[j],
          end: rEnds[j]
        }
        events.push(temp2);
      }
    }
  }
  $('#calendar').fullCalendar('removeEvents');
  $('#calendar').fullCalendar('addEventSource', events);

  document.getElementById("ics_content").value = output;
  // Exporting to file?
  var aasdf = document.getElementById("download_a");
  var file = new Blob([output], {type: 'data:text/ics;charset=utf-8'});
  aasdf.href = URL.createObjectURL(file);
  aasdf.download = "asdf.ics";
  var aasdf = document.getElementById("download_btn");
  aasdf.removeAttribute("disabled");
}

function parseText(input) {
  // Clean up input text
  input = input.split("\n");
  for (var i=input.length-1; i>=0; i--) {
    if (/^\t*$/.exec(input[i])) input.splice(i,1);
  }

  // Iterate through Quest text
  var rows = [];
  for (var i=0; i<input.length; i++) {

    if (result = /(\w+ \d+\w*) - (.+$)/.exec(input[i])) {
      var row = {};
      row["Course Code"] = result[1];
      row["Course Name"] = result[2];
      i++;

      // Skip n lines; this corresponds to the first table, Course Info
      i += input[i].split("\t").length + 1;

      // Get column headings
      var colNames = input[i].split("\t");
      // Trim the headings for trailing spaces
      for (var j=0; j<colNames.length; j++) colNames[j] = colNames[j].trim();
      i++;

      // Distinguish between add'l row in table and end of table (start of Exam Information section)
      while (true) {
        row = readRow(row, input, i, colNames);
        rows.push(row);
        i += colNames.length;
        if (input[i] === "Exam Information") i++;
        if (/^\d{4}$/.exec(input[i+1])) i++;
        if (!/^\d{4}$/.exec(input[i])) break;
      }
      i--;

    }

  }

  return rows;
}

// Produce a struct with fields named with column headers
function readRow(prevRow, input, i, colNames) {
  // make a deep copy of the previous row struct to carry over inferred values
  var row =  {};
  for (var key in prevRow) row[key] = prevRow[key];
    
  // we've already stripped out all lines that are just tab;
  // every other space should be an empty cell
  var n = colNames.length;
  var m = 0;
  for (var j=0; j<n; j++) {
    console.log(i,j);
    // an empty cell implies carry the same data over as previous entry
    if (input[i+j].match(/^[\t\s]*$/)) continue;
    // otherwise just assign take the data and assign it to this field
    row[colNames[m]] = input[i+j];
    if (input[i+j].match(/,\s$/)) {
    row[colNames[m]] = input[i+j] + input[i+j+1];
      j++;
      n++;
    }
    m++;
  }
  console.log(row);
  return row;
}

function populateFormatTable(rows, formatTable) {
  // Determine set of components
  var compSet = new Set();
  for (var i=0; i<rows.length; i++) {
    compSet.add(rows[i]["Component"]);
  }
  // default entry in formatting table
  var compStrs = Array.from(compSet.values()).concat("default");

  // Populate table
  var tr0 = document.createElement("tr");
  var tr1 = document.createElement("tr");
  for (var i=0; i<compStrs.length; i++) {
    var td = document.createElement("td");
    var content = document.createTextNode(compStrs[i].concat(" format"));
    td.appendChild(content);
    tr0.appendChild(td);

    td = document.createElement("td");
    content = document.createElement("input");
    content.id = compStrs[i].concat("_format");
    td.appendChild(content);
    tr1.appendChild(td);
  }
  formatTable.replaceChild(tr0, formatTable.childNodes[0]);
  formatTable.replaceChild(tr1, formatTable.childNodes[1]);

  // Components I've seen include LEC, TUT, LAB, TST, SEM, PRJ
  // LEC and TST are the ones for which I have different specifications; the rest can use default
  // Pre-populate input box for LEC, TST, and default
  var inputField;
  inputField = document.getElementById("LEC_format");
  if (inputField !== null) inputField.value = "%cc - %cn";
  inputField = document.getElementById("TST_format");
  if (inputField !== null) inputField.value = "%cc Midterm";
  document.getElementById("default_format").value = "%cc %comp";
}

function createEvent(rItem) {
	var summary, desc, location, dtstamp, dtstart, dtend, rrule, uid;


  // One of LEC, TUT, LAB, TST, SEM, PRJ, default is used
  // LEC and TST are the ones for which I have different specifications; the rest can use default
  // Summary
  var formatStr = rItem["Component"].concat("_format");
  summary = document.getElementById(formatStr);
  if (summary === null || summary.value === "") { // Either no DOM element returned or the field is empty
    summary = document.getElementById("default_format");
  }
  summary = summary.value;

  // Process substitutions
  summary = summary.split("\\\\");
  for (var i=0; i<summary.length; i++) {
    var temp = summary[i];
    temp = temp.replace(/([^\\]|^)%cc/g, "$1" + rItem["Course Code"]);
    temp = temp.replace(/\\%cc/g, "%cc");
    temp = temp.replace(/([^\\]|^)%cn/g, "$1" + rItem["Course Name"]);
    temp = temp.replace(/\\%cn/g, "%cn");
    temp = temp.replace(/([^\\]|^)%comp/g, "$1" + rItem["Component"]);
    summary[i] = temp.replace(/\\%comp/g, "%comp");
  }
  summary = summary.join("\\");

  // Description
  desc = "Class Nbr: " + rItem["Class Nbr"];
  desc += "\\nSection: " + rItem["Section"];
  desc += "\\nInstructor: " + rItem["Instructor"];

  // Location
  location = rItem["Room"].split(/\s+/).join(" ");

  // DTSTAMP
  // format is YYYMMDDThhmmssZ
  dtstamp = new Date().toISOString().replace(/[-:]/g, "").replace(/\.\d{3}/, "");

  // DTSTART and DTEND and RRULE
  var seSplit = rItem["Start/End Date"].replace(/ - /, '/').split('/');
  // Two different possible date orders
  if (seSplit[0].match(/\d{4}/)) {
    var startDate = seSplit[0] + seSplit[1] + seSplit[2];
    var dayOfWeek = new Date(seSplit[0], seSplit[1]-1, seSplit[2]).getDay();
    var endDate   = seSplit[3] + seSplit[4] + seSplit[5];
  } else {
    var startDate = seSplit[2] + seSplit[0] + seSplit[1];
    var dayOfWeek = new Date(seSplit[2], seSplit[0]-1, seSplit[1]).getDay();
    var endDate   = seSplit[5] + seSplit[3] + seSplit[4];
  }

  var recurring = startDate !== endDate;
  if (recurring) {
    // Regular recurring component
    rrule = createRRule(rItem["Days & Times"], startDate, endDate);
    var temp = createDtBounds(rItem["Days & Times"], startDate, endDate, dayOfWeek, rrule);
    dtstart = temp[0];
    dtend = temp[1];
  } else {
    // Midterm or Seminar or Lab
    var temp = createDtBounds(rItem["Days & Times"], startDate, endDate, dayOfWeek, rrule);
    dtstart = temp[0];
    dtend = temp[1];
  }

  // UID
  uid = dtstamp.replace(/[TZ]/g, "");
  uid += uidCounter;
  uid += "\@aruu.github.io";
  uidCounter++;

  // Sample VEVENT format

  // BEGIN:VEVENT
  // DTSTART;TZID=America/Toronto:20150415T143000
  // DTEND;TZID=America/Toronto:20150415T170000
  // RRULE:FREQ=WEEKLY;UNTIL=20150520T183000Z;BYDAY=WE
  // DTSTAMP:20150415T063712Z
  // UID:a6elgdb5paod2eajn8uehscbe4@google.com
  // CREATED:20150415T063702Z
  // DESCRIPTION:asdf\nasdf\nasdf\nasf
  // LAST-MODIFIED:20150415T063702Z
  // LOCATION:asd  as
  // SEQUENCE:0
  // STATUS:CONFIRMED
  // SUMMARY:asdfasdf
  // TRANSP:OPAQUE
  // END:VEVENT

  // BEGIN:VEVENT
  // DTSTART:20150414T153000Z
  // DTEND:20150414T180000Z
  // DTSTAMP:20150414T023441Z
  // UID:i5hiorj4c5k5jieieo19p8fet0@google.com
  // CREATED:20150414T023411Z
  // DESCRIPTION:desc\nasdf\nasdf
  // LAST-MODIFIED:20150414T023411Z
  // LOCATION:place
  // SEQUENCE:0
  // STATUS:CONFIRMED
  // SUMMARY:title
  // TRANSP:OPAQUE
  // END:VEVENT

  // Print to string
  var outputString = "";
  outputString += "BEGIN:VEVENT\n";
  outputString += "DTSTART;TZID=America/Toronto:" + dtstart + "\n";
  outputString += "DTEND;TZID=America/Toronto:" + dtend + "\n";
  if (recurring) outputString += "RRULE:" + rrule + "\n";
  outputString += "DTSTAMP:" + dtstamp + "\n";
  outputString += "UID:" + uid + "\n";
  // outputString += "CREATED:\n";
  outputString += "DESCRIPTION:" + desc + "\n";
  // outputString += "LAST-MODIFIED:\n";
  outputString += "LOCATION:" + location + "\n";
  // outputString += "SEQUENCE:0\n";
  // outputString += "STATUS:CONFIRMED\n";
  outputString += "SUMMARY:" + summary + "\n";
  // outputString += "TRANSP:OPAQUE\n";
  outputString += "END:VEVENT\n";
  return outputString;
}

function createRRule(daysTimes, startDate, endDate) {
  var rrule = "FREQ=WEEKLY;UNTIL=" + endDate + "T235959Z;BYDAY=";

  var daysChar = daysTimes.match(/\w+/)[0].match(/[MTWF]h?/g);
  var i = 0;
  if (daysChar[i] === "M") {
    rrule += "MO,";
    i++;
  }
  if (daysChar[i] === "T") {
    rrule += "TU,";
    i++;
  }
  if (daysChar[i] === "W") {
    rrule += "WE,";
    i++;
  }
  if (daysChar[i] === "Th") {
    rrule += "TH,";
    i++;
  }
  if (daysChar[i] === "F") {
    rrule += "FR,";
    i++;
  }

  return rrule.replace(/.$/,"");
}

function createDtBounds(daysTimes, startDate, endDate, dayOfWeek, rrule) {

  if (startDate !== endDate) {
    // this is so utterly messy. rewrite how rrules are calculated
    var days = [0,0,0,0,0,0,0];
    if (rrule.match(/BYDAY=.*/)[0].match(/MO/)) days[1] = 1;
    if (rrule.match(/BYDAY=.*/)[0].match(/TU/)) days[2] = 1;
    if (rrule.match(/BYDAY=.*/)[0].match(/WE/)) days[3] = 1;
    if (rrule.match(/BYDAY=.*/)[0].match(/TH/)) days[4] = 1;
    if (rrule.match(/BYDAY=.*/)[0].match(/FR/)) days[5] = 1;

    var offset = 0;
    // increment the start date until it's a valid date.
    while (!days[(dayOfWeek+offset)%7]) offset++;
    startDate = (parseInt(startDate) + offset).toString();
  }

  var pattern = /(\d+):(\d+)(\w+)/g;
  var aData = pattern.exec(daysTimes);
  var bData = pattern.exec(daysTimes);
  if (aData[3] === "PM" && aData[1] !== "12") {
    aData[1] = (parseInt(aData[1]) + 12).toString();
  }
  if (aData[3] === "AM" && aData[1] === "12") {
    aData[1] = "00";
  }
  if (bData[3] === "PM" && bData[1] !== "12") {
    bData[1] = (parseInt(bData[1]) + 12).toString();
  }
  if (bData[3] === "AM" && bData[1] === "12") {
    bData[1] = "00";
  }
  bData[2] = (parseInt(bData[2]) + 10).toString();
  if (bData[2] === "60") {
    bData[1] = (parseInt(bData[1]) + 1).toString();
    bData[2] = "00";
  }
  
  var a = aData[1] + aData[2] + "00";
  for (var i=a.length; i<6; i++) a = "0" + a;
  var b = bData[1] + bData[2] + "00";
  for (var i=b.length; i<6; i++) b = "0" + b;

  return [startDate+"T"+a, startDate+"T"+b];
}
